### PR TITLE
Add new zally statistics

### LINKED ISSUE #360: Zally Statistics - Further Requirements
**As** Zally user and/or API Guild member
**I want to** have meaningful statistics in place
**So that** information about the usage and rule violations is available

After some discussions with Product, we collected the requirements for how the reports should look like. We also decided on the way we want to implement this feature. The discussion, outcome and some details are stated in [this internal document](https://docs.google.com/document/d/1yOG96GwnQ9NFWPC5Zutby_HDJ0crYk-rM4DfybreDYY/edit).

Part of the requirements is already implemented in #320.

### PR DESCRIPTION
**Pull request by fmueller**
Closes #360 and #356

### CODE REVIEW

File: server/src/main/java/de/zalando/zally/apireview/ApiReview.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        this.successfulProcessed = StringUtils.isNotBlank(apiDefinition);
+        this.created = Instant.now().atOffset(ZoneOffset.UTC);
+        this.day = created.toLocalDate();
+
+        this.ruleViolations = violations.stream()
+            .map(v -> new RuleViolation(this, v.getRule().getName(), v.getViolationType(), v.getPaths().size()))
+            .collect(Collectors.toList());
+
+        this.numberOfEndpoints = EndpointCounter.count(apiDefinition);
+        this.mustViolations = (int) ruleViolations.stream().filter(r -> r.getType() == ViolationType.MUST).count();
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/apireview/ApiReview.java]: `(int) ruleViolations.stream().filter(r -> r.getType() == ViolationType.MUST).count()` can be a nice helper function

File: server/src/main/java/de/zalando/zally/apireview/EndpointCounter.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+public final class EndpointCounter {
+
+    private EndpointCounter() {
+    }
+
+    public static int count(String apiDefinition) {
+        try {
+            Swagger swagger = new SwaggerParser().parse(apiDefinition);
+            return swagger != null ? swagger.getPaths().size() : 0;
+        } catch (Exception e) {
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/apireview/EndpointCounter.java]: Is it possible to catch a precise exception here?

File: server/src/main/java/de/zalando/zally/configuration/JacksonObjectMapperConfiguration.java [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
     public ObjectMapper createObjectMapper() {
-        return new ObjectMapper().registerModules(
-            new Jdk8Module(),
-            new JavaTimeModule(),
-            new ProblemModule(),
-            new KotlinModule());
+        return new ObjectMapper()
+            .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
+            .configure(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS, false)
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/configuration/JacksonObjectMapperConfiguration.java]: Nice!

File: server/src/main/resources/db/migration/V001__add_base_schema.sql [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
+  name          TEXT NOT NULL,
+  type          VARCHAR(10) NOT NULL,
+  occurrence    INT NOT NULL
 );
 
-ALTER TABLE review_statistic ADD CONSTRAINT unique_value_per_day UNIQUE (name, value, day);
-
-CREATE INDEX api_review_request_created_idx ON api_review_request (created);
-CREATE INDEX report_statistic_day_idx ON review_statistic (day);
+CREATE INDEX api_review_day_idx ON api_review (day);
```
Comment (netme) [inline server/src/main/resources/db/migration/V001__add_base_schema.sql]: Do we have any data to migrate? Most probably no, just wanted to cross-check.

**Comment by netme** [comment]
👍

File: server/src/main/java/de/zalando/zally/apireview/ApiReview.java (hunk 1 above)
Comment (fmueller) [inline server/src/main/java/de/zalando/zally/apireview/ApiReview.java]: Will change it.

File: server/src/main/java/de/zalando/zally/apireview/EndpointCounter.java (hunk 2 above)
Comment (fmueller) [inline server/src/main/java/de/zalando/zally/apireview/EndpointCounter.java]: Unfortunately not, the parser throws Exception if something went wrong. In this case it's ok to catch Exception I think as a default value is returned.

File: server/src/main/resources/db/migration/V001__add_base_schema.sql (hunk 4 above)
Comment (fmueller) [inline server/src/main/resources/db/migration/V001__add_base_schema.sql]: No, database is not yet deployed. We are safe. :)

**Comment by fmueller** [comment]
@netme @maxim-tschumak Incorporated your feedback. Please re-check.

**Comment by netme** [comment]
@fmueller , many thanks, looks good

**Comment by netme** [comment]
👍

**Comment by netme** [comment]
👍

**Comment by maxim-tschumak** [comment]
:+1:

**Pull request merged**